### DISCUSSION TITLE
How do i avoid desync while making a rhythm game?

### DISCUSSION THREAD

**Discussion by Gaypatapon** [TurboWarp/extensions#1125] [Q&A] [answered]
I'm making a patapon fangame but the beat desyncs, i tried using the timer but it still doesn't work, is there a way to make precise rhythm?

**Comment by supersonico** [comment]
I suggest using the DeltaTime extension

**Reply by Gaypatapon** [reply]
ok
I'll try

**Reply by Gaypatapon** [reply]
how do i do it then?

**Reply by softedco** [reply]
delta time isn't related to this

**Reply by Gaypatapon** [reply]
well then how do I avoid desync without needing to have 60 fps?

**Reply by supersonico** [reply]
> delta time isn't related to this

DeltaTime is used in most rythm games to ensure a consistent speed and sync with the songs regardless of the frame rate, how is it not related to this?

**Reply by supersonico** [reply, marked as answer]
> how do i do it then?

Well, there's multiple ways to use DeltaTime, I personally use this:
![image](https://github.com/TurboWarp/extensions/assets/126848249/2f740927-d16a-4b7e-aaf4-a0dbdc5acc35)

I'm sure your favorite browser can help you find more effective ways.

**Reply by Gaypatapon** [reply]
Ok cool, is there also a way to do this with timers?

**Reply by Gaypatapon** [reply]
like I don't want the beat to desync

**Reply by supersonico** [reply]
> Ok cool, is there also a way to do this with timers?

If you mean using DeltaTime with the timer instead of the extension, yes:
https://scratch.mit.edu/projects/487694716/

**Reply by Gaypatapon** [reply]
I meant as in using the timer with the extention

**Reply by Gaypatapon** [reply]
could you show me how to use deltatime for stopping from a rhythm going offsync

**Reply by supersonico** [reply]
> I meant as in using the timer with the extention
You don't necessarily need to use the timer if you use the extension.

**Reply by supersonico** [reply]
> could you show me how to use deltatime for stopping from a rhythm going offsync

I don't know how your rhythm algorithm works, this is just an example of what you should do with your scripts: https://github.com/TurboWarp/extensions/discussions/1125#discussioncomment-7465989

You can find more info about Delta Time and when to use it here: https://en.wikipedia.org/wiki/Delta_timing

**Reply by Gaypatapon** [reply]
Oh, can you show me how to use deltatime for rhythm games?

**Reply by Gaypatapon** [reply]
the rhythm is like from patapon, i'm making a patapon fangame but the problem is that the beat doesn't match up the song

**Reply by supersonico** [reply]
> Oh, can you show me how to use deltatime for rhythm games?

> the rhythm is like from patapon, i'm making a patapon fangame but the problem is that the beat doesn't match up the song

Again, I don't know how your game works nor do I want to know. I already showed you the basic script and an example, you should try to apply it to your scripts.

**Reply by Gaypatapon** [reply]
ok, sorry if i ask too much

**Reply by softedco** [reply]
delta time * fps is 1

**Reply by supersonico** [reply]
> Well, there's multiple ways to use DeltaTime, I personally use this:
 ![image](https://user-images.githubusercontent.com/126848249/280259065-2f740927-d16a-4b7e-aaf4-a0dbdc5acc35.png)

Note that with "your fps" I am not referring to the current FPS, I am referring to the project framerate set from TurboWarp. It's just so you don't have to worry about finding new values for all your scripts when using Delta Time. So my bad there, I should have said framerate instead of fps. 

![image](https://github.com/TurboWarp/extensions/assets/126848249/6ddd23eb-5083-4718-aca8-5e6706841f75)

If this is disabled you should just put 30 in "your fps"

![image](https://github.com/TurboWarp/extensions/assets/126848249/2a83a6fe-c6dd-4758-8bf3-0348fa254d56)

**Comment by softedco** [comment]
> > how do i do it then?
> 
> Well, there's multiple ways to use DeltaTime, I personally use this: ![image](https://user-images.githubusercontent.com/126848249/280259065-2f740927-d16a-4b7e-aaf4-a0dbdc5acc35.png)
> 
> I'm sure your favorite browser can help you find more effective ways.



> > delta time isn't related to this
> 
> DeltaTime is used in most rythm games to ensure a consistent speed and sync with the songs regardless of the frame rate, how is it not related to this?

because it is used in all decent games to ensure sync, if it's used in rhythm games it does not mean it is related to making rhythm sync, it is only related to making the whole project sync, and sync that is with only other framerates, not the music

**Reply by supersonico** [reply]
> because it is used in all decent games to ensure sync, if it's used in rhythm games it does not mean it is related to making rhythm sync, it is only related to making the whole project sync, and sync that is with only other framerates, not the music

I never said Delta Time was exclusively for rhythm games, I said rhythm games use it.
Of course DT is not only for the music, but the desync usually occurs because of inconsistent FPS, so using DeltaTime is one step closer to avoid desync.

**Reply by softedco** [reply]
It doesn't usually occur when you have a fixed framerate which is true for almost all games made in scratch and turbowarp

**Reply by supersonico** [reply]
Even if you set the framerate to a value like 30, 60, 250 (as TurboWarp allows you to do), it does not mean that the FPS will always be that, you just set the max FPS.

**Comment by softedco** [comment]
> > Oh, can you show me how to use deltatime for rhythm games?
> 
> > the rhythm is like from patapon, i'm making a patapon fangame but the problem is that the beat doesn't match up the song
> 
> Again, I don't know how your game works nor do I want to know. I already showed you the basic script and an example, you should try to apply it to your scripts.

You didn't answer him on how to make music sync with gameplay, which is what he was asking in the first place.

**Reply by supersonico** [reply]
They were asking how to avoid desync in a rhythm game, read the title.

**Reply by softedco** [reply]
it's about beat desync, not game desync

**Comment by GarboMuffin** [comment]
Delta time isn't really meant to solve this. The best you're going to get is using "days since 2000" to get the most precise timer available in Scratch (multiply by 86400 to get seconds; you could also just use normal timer blocks). Record the start time and just calculate the current time in the song from that.

Instead of doing things like "move the note forward [x] pixels per frame" (will lead to imprecision), you would use some math to figure out exactly where each note is supposed to be displayed at the current time. That's the best you can get.

**Reply by Skyhigh173** [reply]
yep I use this method on my rhythm game
